File: Python/ExampleTests/common/order_item_string.py
from . import tokens
DNC_STRING = '?DNC?'


class OrderItemString:
    def __init__(self, name: str = '', quantity: str = '', price: str = '', item_total: str = ''):
        self.name = name
        self.quantity = quantity
        self.price = price
        self.item_total = item_total

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else '',
            v[3] if len(v) > 3 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 4, 'OrderItem')
        return cls(
            parts[0],
            parts[1],
            parts[2],
            parts[3]
        )

    def __str__(self):
        return (tokens.token(self.name) + ' ' +
                tokens.token(self.quantity) + ' ' +
                tokens.token(self.price) + ' ' +
                tokens.token(self.item_total))

    def _key(self):
        return (self.name, self.quantity, self.price, self.item_total)

    def __eq__(self, other):
        if not isinstance(other, OrderItemString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'OrderItemString({self})'

File: Swift/ExampleTests/common/OrderItemString.swift
public struct OrderItemString: CustomStringConvertible, Equatable {
    public let name: String
    public let quantity: String
    public let price: String
    public let itemTotal: String

    public init(name: String, quantity: String, price: String, itemTotal: String) {
        self.name = name
        self.quantity = quantity
        self.price = price
        self.itemTotal = itemTotal
    }

    public init(fromArray v: [String]) {
        self.name = v.count > 0 ? v[0] : ""
        self.quantity = v.count > 1 ? v[1] : ""
        self.price = v.count > 2 ? v[2] : ""
        self.itemTotal = v.count > 3 ? v[3] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> OrderItemString {
        let parts = Tokens.require(text, 4, "OrderItem")
        return OrderItemString(name: parts[0], quantity: parts[1], price: parts[2], itemTotal: parts[3])
    }

    public var description: String {
        return Tokens.token(name) + " " + Tokens.token(quantity) + " " + Tokens.token(price) + " " + Tokens.token(itemTotal)
    }

    public static let dncString = "?DNC?"

    public static func == (a: OrderItemString, b: OrderItemString) -> Bool {
        return (a.name == dncString || b.name == dncString || a.name == b.name)
            && (a.quantity == dncString || b.quantity == dncString || a.quantity == b.quantity)
            && (a.price == dncString || b.price == dncString || a.price == b.price)
            && (a.itemTotal == dncString || b.itemTotal == dncString || a.itemTotal == b.itemTotal)
    }
}
